### PR TITLE
separate out our Tokenizer

### PR DESCRIPTION
**Pull request by karpathy**

### CODE REVIEW

File: tokenizer.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
+/*
+Defines the GPT-2 Tokenizer.
+Only supports decoding, i.e.: tokens (integers) -> strings
```
Comment (ngc92) [inline tokenizer.h:3]: Only supports **decoding**, i.e.: tokens (integers) -> strings
This is all we need for unconditional generation.
If we wanted to later prompt the model, we'd have to add **decoding**  :)

Comment (karpathy) [reply tokenizer.h:3]: hehe oops

**Comment by karpathy** [comment]
I did better than just move it, I am now saving it in the Tokenizer file in the header, which is a lot better.

File: train_gpt2.py [hunk 2]
```diff
@@ -351,8 +351,9 @@ def write_tokenizer(enc, filename):
     n = enc.max_token_value + 1
     header = torch.zeros(256, dtype=torch.int32)
     header[0] = 20240328 # magic
-    header[1] = 1 # tokenizer version = 1
+    header[1] = 2 # tokenizer version = 2 (1 -> 2: includes EOT token)
     header[2] = n # number of tokens
+    header[3] = enc.eot_token # EOT token
```
Comment (azret) [inline train_gpt2.py:356]: Yep. Perfect! Thank you.

**Comment by azret** [comment]
What about the C/CPU version?

**Pull request merged**